src/main/java/org/broadinstitute/hellbender/engine/dataflow/datasources/ShardedReadsDatasource.java

```diff
@@ -0,0 +1,361 @@
+        List<KV<String,Void>> intervalKeys;
+        intervalKeys = shards.stream().map(s-> KV.<String,Void>of(s.toString(), null)).collect(Collectors.<KV<String,Void>>toList());
+
+        return pipeline
+            // ship shards to workers
+            .apply(Create.<SimpleInterval>of(shards)).setName("shards")
+                // for each shard, load the corresponding reads
+            .apply(ParDo.named("getGroupedReadPCollection").of(new DoFnWLog<SimpleInterval, KV<String, Iterable<GATKRead>>>("getGroupedReadPCollection") {
+                private static final long serialVersionUID = 1L;
+                private transient Storage.Objects storageClient;
+                private transient Stopwatch querying;
+                private transient Stopwatch loading;
+                private transient Stopwatch choosing;
+                private transient Stopwatch shipping;
+                private long readsEmitted = 0;
+                Date started;
+
+                @Override
+                public void startBundle(Context c) throws Exception {
+                    super.startBundle(c);
+                    if (cloudStorageUrl) {
+                        storageClient = GCSOptions.Methods.createStorageClient(c.getPipelineOptions().as(GCSOptions.class), auth);
+                    }
+                    started = new Date();
+                    readsEmitted = 0;
+                    querying = Stopwatch.createUnstarted();
+                    loading = Stopwatch.createUnstarted();
+                    choosing = Stopwatch.createUnstarted();
+                    shipping = Stopwatch.createUnstarted();
+                }
+
+                @Override
+                public void finishBundle(Context c) throws Exception {
+                    super.finishBundle(c);
+                    final Instant instant = new Date().toInstant();
+                    long ms = Duration.between(started.toInstant(), instant).toMillis();
+                    String q = querying.elapsed(TimeUnit.MILLISECONDS) + " ms";
+                    String l = loading.elapsed(TimeUnit.MILLISECONDS) + " ms";
+                    String ch = choosing.elapsed(TimeUnit.MILLISECONDS) + " ms";
+                    String s = shipping.elapsed(TimeUnit.MILLISECONDS) + " ms";
+                    bunny.stepEnd("Bundle sent " + readsEmitted + " reads in " + ms + " ms. Breakdown: querying " + q + " loading " + l + " choosing " + ch + " shipping " + s);
+                }
+
+                @Override
+                public void processElement(ProcessContext c) throws Exception {
+                    int rejected = 0;
+                    int accepted = 0;
+                    Stopwatch processing = Stopwatch.createStarted();
+
+                    Stopwatch kving = Stopwatch.createUnstarted();
+                    Stopwatch outputting = Stopwatch.createUnstarted();
+
+                    Stopwatch readingInput = Stopwatch.createStarted();
```

Comment:
Do we really want all these `Stopwatch` start/stop statements in the master branch? The code is almost unreadable like this...

If they are still needed for now, perhaps we could have both a clean version and a version for profiling?